flink-java/src/main/java/org/apache/flink/api/java/functions/AssignRangeIndex.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.java.functions;
+
+import org.apache.flink.api.common.distributions.DataDistribution;
+import org.apache.flink.api.common.functions.RichMapPartitionFunction;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.Collector;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This mapPartition function require a DataSet with DataDistribution as broadcast input, it read
+ * target parallelism from parameter, build partition boundaries with input DataDistribution, then
+ * compute the range index for each record.
+ *
+ * @param <IN> The original data type.
+ * @param <K> The key type.
+ */
+public class AssignRangeIndex<IN, K extends Comparable<K>>
+	extends RichMapPartitionFunction<Tuple2<K, IN>, Tuple2<Integer, IN>> {
+
+	private List<K> partitionBoundaries;
+	private int numberChannels;
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		this.numberChannels = parameters.getInteger("TargetParallelism", 1);
+	}
+
+	@Override
+	public void mapPartition(Iterable<Tuple2<K, IN>> values, Collector<Tuple2<Integer, IN>> out) throws Exception {
+
+		List<Object> broadcastVariable = getRuntimeContext().getBroadcastVariable("DataDistribution");
+		if (broadcastVariable == null || broadcastVariable.size() != 1) {
```

Comment:
Nevermind, I thought you were using a MapFunction, but its a MapPartitionFunction. So this is only done once.